#### Tools

1.Custom funtions

2. Built In

3. 3rd party

In [1]:
import asyncio
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from autogen_core.tools import FunctionTool
import os
from dotenv import load_dotenv

In [2]:
load_dotenv()
api_key = os.getenv("OPENAI_API_KEY")
if not api_key:
    raise ValueError("please set the OPENAI_API_KEY env varible")

In [3]:
model_client = OpenAIChatCompletionClient(model="gpt-4o", api_key=api_key)

In [4]:
def reverse_string(text:str) -> str:
    '''
    Reverse the given text

    input :str
    output:str
    the reversed string is returned
    '''

    return text[::-1]

In [5]:
reverse_tool = FunctionTool(reverse_string, description='a tool to reverse a string')

In [ ]:
agent = AssistantAgent(
    name="ReverseStringAgent",
    model_client=model_client,
    system_message="you are a helpful assistant that can reverse string using reverse_string tool. Give the result with sumary",
    tools=[reverse_tool]
)

In [8]:
result = await agent.run(task='reverse the string "Hello World!"')

In [9]:
result

TaskResult(messages=[TextMessage(id='96a26eb4-fea6-4221-acc9-3b6a86d2553e', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 9, 4, 1, 43, 46, 390139, tzinfo=datetime.timezone.utc), content='reverse the string "Hello World!"', type='TextMessage'), ToolCallRequestEvent(id='469fa72b-1003-4543-960e-12d6155e8576', source='ReverseStringAgent', models_usage=RequestUsage(prompt_tokens=79, completion_tokens=16), metadata={}, created_at=datetime.datetime(2025, 9, 4, 1, 43, 48, 56154, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='call_ahmy4VdyveSsQ3vFXPBKqJEd', arguments='{"text":"Hello World!"}', name='reverse_string')], type='ToolCallRequestEvent'), ToolCallExecutionEvent(id='0d480e5d-c3b4-47ba-9eaa-94bcbbfe579a', source='ReverseStringAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 9, 4, 1, 43, 48, 58399, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content='!dlroW olleH', name='reverse_string', call_id='c

TaskResult(messages=[TextMessage(id='96a26eb4-fea6-4221-acc9-3b6a86d2553e', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 9, 4, 1, 43, 46, 390139, tzinfo=datetime.timezone.utc), content='reverse the string "Hello World!"', type='TextMessage'), ToolCallRequestEvent(id='469fa72b-1003-4543-960e-12d6155e8576', source='ReverseStringAgent', models_usage=RequestUsage(prompt_tokens=79, completion_tokens=16), metadata={}, created_at=datetime.datetime(2025, 9, 4, 1, 43, 48, 56154, tzinfo=datetime.timezone.utc), content=[FunctionCall(id='call_ahmy4VdyveSsQ3vFXPBKqJEd', arguments='{"text":"Hello World!"}', name='reverse_string')], type='ToolCallRequestEvent'), ToolCallExecutionEvent(id='0d480e5d-c3b4-47ba-9eaa-94bcbbfe579a', source='ReverseStringAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 9, 4, 1, 43, 48, 58399, tzinfo=datetime.timezone.utc), content=[FunctionExecutionResult(content='!dlroW olleH', name='reverse_string', call_id='call_ahmy4VdyveSsQ3vFXPBKqJEd', is_error=False)], type='ToolCallExecutionEvent'), ToolCallSummaryMessage(id='ea5d9775-94dc-45bb-bcb5-ef10f92ffbb3', source='ReverseStringAgent', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 9, 4, 1, 43, 48, 58465, tzinfo=datetime.timezone.utc), content='!dlroW olleH', type='ToolCallSummaryMessage', tool_calls=[FunctionCall(id='call_ahmy4VdyveSsQ3vFXPBKqJEd', arguments='{"text":"Hello World!"}', name='reverse_string')], results=[FunctionExecutionResult(content='!dlroW olleH', name='reverse_string', call_id='call_ahmy4VdyveSsQ3vFXPBKqJEd', is_error=False)])], stop_reason=None)

In [12]:
result.messages[-1].content


'!dlroW olleH'